### PR TITLE
Fixes #25614 - v2 Features api with Capabilities and exposed settings

### PR DESCRIPTION
**Pull request by jlsherrill**

### CODE REVIEW

**Comment by jlsherrill** [comment]
This is a take over of https://github.com/theforeman/smart-proxy/pull/585 (with his blessing).  It maintains most all of his original pr, but:

1) Adds authentication as a requirement to the v2 features api
2) Allows capabilities to be calculated dynamically by simply defining them with a proc:

```
capability(Proc.new { 
   calculate_something_here
})
```

File: modules/root/protected_root_api.rb
Comment (ekohl) [inline modules/root/protected_root_api.rb]: Was this part of debugging?

File: modules/root/protected_root_api.rb
Comment (ekohl) [inline modules/root/protected_root_api.rb]: I like the ability of using a proc. I didn't think of that. Should we take some safe guards here with regards to exceptions or timeouts (if they call to external services)?

A possible test case in the current code would be to use puppet. If it has a provider set, then it should have the capability puppetrun. This can then be used by the foreman to detect it at runtime. Currently it's a setting in foreman's settings.yaml to enable the button globally but it's really a capability on the proxy for the puppet feature.

File: modules/root/protected_root_api.rb
Comment (ekohl) [inline modules/root/protected_root_api.rb]: Should this be an internal server error?

File: modules/root/protected_root_api.rb
Comment (ekohl) [inline modules/root/protected_root_api.rb]: Later I realized that sending the state can be useful. It would mean the server could detect misconfigurations (which is now done in the installer). That does mean the server must always filter on running plugins so it'd be good to get it in on the first go.

**Review by ekohl — COMMENTED** [review]
Thanks for taking over the effort. It looks like I didn't add tests yet either.

File: modules/root/protected_root_api.rb
Comment (jlsherrill) [reply modules/root/protected_root_api.rb]: Whoops yes, will uncomment :)

File: modules/root/protected_root_api.rb
Comment (jlsherrill) [reply modules/root/protected_root_api.rb]: Yeah, after discussing with eric, i realized that the installer may not even know what content units are installed (for pulp 3), so dynamic capabilities became more important.  I went back and forth between explicitly dynamic properties, but this seemed very simple and worked well. 

I'm not sure that we should enforce a timeout outside of the proc call, RestClient will typically timeout after ~2 minutes, and i think it's simpler to leave it up to the plugin to ensure that a timeout will occur.  Otherwise the only other way i can think would be to spin up the proc in a thread and kill it after some timeout occurs.  I'm not sure thats worth it at this time, but open to other opinions.

Comment (ekohl) [reply modules/root/protected_root_api.rb]: I'd be OK with mentioning in docs that you are expected to handle all exceptions and timeouts.

File: modules/root/protected_root_api.rb
Comment (jlsherrill) [reply modules/root/protected_root_api.rb]: This was code you wrote :) but yes i think you are correct!

Comment (ekohl) [reply modules/root/protected_root_api.rb]: I thought I could pin it on you now ;) But yes, my was very much WIP

File: modules/root/protected_root_api.rb
Comment (jlsherrill) [reply modules/root/protected_root_api.rb]: That makes sense.  I think we'd want to ignore dynamic capabilities if its not running?

Comment (ekohl) [reply modules/root/protected_root_api.rb]: Good thought

**Comment by jturel** [comment]
After I click 'Refresh features' in the UI I see that the version API no longer works and the UI reports 
```
ERF50-5345 [Foreman::WrappedException]: Unable to connect ([ProxyAPI::ProxyException]: ERF12-1764 [ProxyAPI::ProxyException]: Unable to detect version ([RestClient::NotFound]: 404 Not Found) for proxy https://centos7-proxy-devel.jturel.example.com:9090/version)
```
With a 404 in the proxy logs:

```
2019-01-07T19:02:55 492926e0 [I] Started GET /version 
2019-01-07T19:02:55 492926e0 [I] Finished GET /version with 404 (0.77 ms)
```

However, it version check seems to work *before* I refresh features even running the new code

**Comment by jlsherrill** [comment]
@jturel that's strange, my code shouldn't have affected that at all.  Any errors in the smart proxy log?

**Comment by jturel** [comment]
@jlsherrill it could very well be a PICNIC. I didn't see any docs on how to test this so what I did was:

- spin up centos7-proxy-devel
- cd to /usr/share/foreman-proxy
- git init to preserve state
- cherry pick the commits on this PR
- restart foreman proxy

Is there a better way?

edit: no errors in the proxy log though - just the 404

**Comment by jlsherrill** [comment]
@jturel yeah, that should be fine, although i actually tested running straight from git (which was a little bit more work).

To clarify, fetching the version worked until you clicked refresh?  and now it doesn't work?  Did you try curling it directly?

**Comment by jturel** [comment]
When I switch to the branch with your changes upon first load of the UI for my external proxy I see that it actually doesn't fetch the version. I guess it's cached (hence refresh button). So .. ignore what I said about clicking refresh!

When I use the foreman API I see the error as well:

https://centos7-devel.jturel.example.com/api/v2/smart_proxies/2/version
```
{
"error": {
"message": "ERF50-5345 [Foreman::WrappedException]: Unable to connect ([ProxyAPI::ProxyException]: ERF12-1764 [ProxyAPI::ProxyException]: Unable to detect version ([RestClient::NotFound]: 404 Not Found) for proxy https://centos7-proxy-devel.jturel.example.com:9090/version)"
}
}
```
And the same 404 in the proxy logs:
```
2019-01-07T19:42:19 492926e0 [I] Started GET /version
2019-01-07T19:42:19 492926e0 [I] Finished GET /version with 404 (0.39 ms)
```

I guess you aren't seeing this problem?

**Comment by jlsherrill** [comment]
@jturel good catch, updated with a new commit

**Comment by jlsherrill** [comment]
[test core]

**Comment by jturel** [comment]
@jlsherrill works now! could you enumerate some of the overall goals of these two PRs so I can be a little more targeted in my testing? Existing functionality seems to be met just fine

**Comment by jlsherrill** [comment]
@jturel  See: https://community.theforeman.org/t/exposing-capabilities-in-the-smart-proxy/9860/1

Its kinda hard to test fully as not much utilizes it, the only thing actually utilized in this pr is the tftpserver name setting.  

What you can do is edit an enabled smart proxy plugin, such as puppet on here:
https://github.com/theforeman/smart-proxy/blob/develop/modules/puppet_proxy/puppet_plugin.rb#L7

on that line add:
```
expose_setting :puppet_version:

capability('static_capability')
capability(Proc.new {
  ['some', 'dynamic', 'capabilities']
})

```

This will expose a setting, a static capability, and a dynamic capability. 

and then do a refresh and check the SmartProxyFeatures model, specifically SmartProxyFeature#capabilities and SmartProxyFeature#settings.  You might try changing around the capabilities and settings, do a refresh, and verify that things get updated in the db.

**Comment by jlsherrill** [comment]
I forgot to mention, you need to restart the smart proxy after any change.

File: test/plugins/plugin_test.rb
Comment (jturel) [inline test/plugins/plugin_test.rb:133]: Can you add a cabability test which uses a lambda as well?

I'm getting an error once I restart my proxy when defining a capability like this:
```
capability(lambda{ ['a', 'b'] })
```

```
Error during startup, terminating. comparison of String with Proc failed
```

I think your tests in root_v2_api_tests didn't reveal it because of the way you're stubbing the `loaded` method but I'm just guessing

**Comment by jturel** [comment]
Thanks! This is meeting the design laid out by the RFC and I see that `:tftp_servername` made it into my DB correctly. I left one other comment regarding testing.

**Comment by jlsherrill** [comment]
@jturel fixed, we were previously sorting capabilities at plugin registration time, this was moved to feature request time to enure that dynamic capabilities were sorted properly as well.  Also test was added to catch that

**Comment by jlsherrill** [comment]
@lzap the failing test seems to also occur on master?

File: lib/proxy/plugin_initializer.rb
Comment (ekohl) [inline lib/proxy/plugin_initializer.rb]: I probably wrote this initially, but could this be `members.map(&:capabilities).compact.flatten`?

File: modules/root/root_v2_api.rb
Comment (jturel) [inline modules/root/root_v2_api.rb:11]: Interesting. Is there a case where there'd be a plugin that would not inherit ::Proxy::Plugin, or is this more of a defensive check?

Comment (ekohl) [reply modules/root/root_v2_api.rb:11]: I copied it from the v1 api:
https://github.com/theforeman/smart-proxy/blob/ce744b8bee2506fe8779f561ded3977dc15feabb/modules/root/root_api.rb#L6

Comment (jlsherrill) [reply modules/root/root_v2_api.rb:11]: This is actually copied from https://github.com/theforeman/smart-proxy/blob/develop/modules/root/root_api.rb#L6   in @ekohl's original PR.  I don't actually know the reason for it :)

File: modules/root/root_v2_api.rb
Comment (jturel) [inline modules/root/root_v2_api.rb:31]: If I had written this I'd have checked for `respond_to? :call`

Not asking for a change but wanted to mention it in case you agree since I see checking for Proc in a few places

**Review by jturel — APPROVED** [review]
I'm acking because it works to the best of my knowledge, but probably better for someone with more smart-proxy background to ACK as well. I just wanted to get more familiar with this stuff :)

File: modules/root/root_v2_api.rb
Comment (jturel) [reply modules/root/root_v2_api.rb:11]: Looks like it was https://projects.theforeman.org/issues/15025 I guess it's fair enough to keep

**Comment by jturel** [comment]
Oh - I think https://projects.theforeman.org/projects/smart-proxy/wiki/API will need some updates as well :)

**Comment by jlsherrill** [comment]
@jturel yeah, i will update that once merged

**Comment by lzap** [comment]
Yes I can confirm that this is a failed test in develop branch, regression. Will look into it later.

**Comment by lzap** [comment]
The fix is here @jlsherrill https://github.com/theforeman/smart-proxy/pull/628

**Comment by jlsherrill** [comment]
Thanks @lzap 
[test core]

File: test/plugins/plugin_initializer_test.rb
Comment (lzap) [inline test/plugins/plugin_initializer_test.rb:19]: Can you enable this plugin? Seems disabled.

File: test/plugins/plugin_test.rb
Comment (lzap) [reply test/plugins/plugin_test.rb:133]: Won't hurt.

File: lib/proxy/pluggable.rb
Comment (lzap) [inline lib/proxy/pluggable.rb:40]: Method not covered in tests.

File: lib/proxy/pluggable.rb
Comment (lzap) [inline lib/proxy/pluggable.rb:188]: Ditto.

**Review by lzap — CHANGES_REQUESTED** [review]
I am okay with this, the major concern are tests now. Once rebased I am gonna try this.

File: test/plugins/plugin_test.rb
Comment (jlsherrill) [reply test/plugins/plugin_test.rb:133]: this was added

**Comment by jlsherrill** [comment]
[test core]

**Comment by jlsherrill** [comment]
[test rubocop]

**Comment by jlsherrill** [comment]
[test core]

**Review by lzap — APPROVED** [review]
WFM, @ekohl any comments?

The output is a bit lengthy, I wonder if we could drop empty arrays/hashes but not a big deal:

```json
{
    "discovery": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "discovery_image": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "salt": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "abrt": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "pulp": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "pulpnode": {
        "http_enabled": false,
        "https_enabled": false,
        "settings": {

        },
        "state": "disabled",
        "capabilities": [

        ]
    },
    "tail": {
        "http_enabled": false,
        "https_enabled": false,
        "settings": {

        },
        "state": "disabled",
        "capabilities": [

        ]
    },
    "dynflow": {
        "http_enabled": false,
        "https_enabled": false,
        "settings": {

        },
        "state": "disabled",
        "capabilities": [

        ]
    },
    "facts": {
        "http_enabled": false,
        "https_enabled": false,
        "settings": {

        },
        "state": "disabled",
        "capabilities": [

        ]
    },
    "dns": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "templates": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "tftp": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {
            "tftp_servername": "192.168.100.1"
        },
        "state": "running",
        "capabilities": [

        ]
    },
    "dhcp": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "puppetca": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "puppet": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "bmc": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "realm": {
        "http_enabled": false,
        "https_enabled": false,
        "settings": {

        },
        "state": "disabled",
        "capabilities": [

        ]
    },
    "logs": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    },
    "httpboot": {
        "http_enabled": true,
        "https_enabled": true,
        "settings": {

        },
        "state": "running",
        "capabilities": [

        ]
    }
}
```

**Review by ekohl — APPROVED** [review]
> WFM, @ekohl any comments?

Yay! Big thanks to @jlsherrill for finish what I started. 

> The output is a bit lengthy, I wonder if we could drop empty arrays/hashes but not a big deal:

In my experience an API is easier if you can assume data structures. Otherwise you need to check if a key is present or a value is nil. IMHO it's worth the added overhead.

**Comment by lzap** [comment]
It's in! Thanks @jlsherrill and @ekohl and others. Let's start building on top of that.

**Comment by lzap** [comment]
This is demo-worthy I think even if Foreman part is not yet finished.

**Pull request merged**
